Bluetooth: controller: Add reset of per adv for HCI_Reset command

ppryga-nordic opened the pull request:
There is missing handling of periodic advertising reset for
HCI_Reset command. That makes impossible to execute e.g.
qualification tests without rebooting of the DUT
for periodic advertising and direction finding.

This commit adds missing implementation.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: Add reset of per adv for HCI_Reset command

There is missing handling of periodic advertising reset for
HCI_Reset command. That makes impossible to execute e.g.
qualification tests without rebooting of the DUT
for periodic advertising and direction finding.

This commit adds missing implementation.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1506,6 +1507,13 @@ int ull_adv_reset_finalize(void)
 	uint8_t handle;
 	int err;
 
+#if defined(CONFIG_BT_CTLR_ADV_PERIODIC)
+	err = ull_adv_sync_reset_finalize();
+	if (err) {
+		return err;
+	}
+#endif /* CONFIG_BT_CTLR_ADV_PERIODIC */
+
```
`_finalise()` is not required, as lll_adv_data_reset for periodic advertising is called on every `ll_adv_sync_param_set` call.

```suggestion
```

ppryga replied:
It seems to me that it is required to clear `adv->lll.sync`.

ppryga replied:
The code is left here.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv_internal.h:
```diff
@@ -150,6 +150,7 @@ void ull_adv_aux_ptr_fill(uint8_t **dptr, uint8_t phy_s);
 
 int ull_adv_sync_init(void);
 int ull_adv_sync_reset(void);
+int ull_adv_sync_reset_finalize(void);
```
```suggestion
```

ppryga replied:
Function not removed.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -635,8 +635,62 @@ int ull_adv_sync_init(void)
+		err = sync_remove(sync, adv, 0U);
+		if (err) {
+			return err;
+		}
+	}
```
```suggestion
		}
		
		adv->lll.sync = NULL;
	}
```

ppryga replied:
`adv->lll.sync = NULL` moved to `ull_adv_reset_finalize` following pattern with `adv->lll.aux=NULL`.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -635,8 +635,62 @@ int ull_adv_sync_init(void)
+			return err;
+		}
+	}
+
+	return 0;
+}
+
+int ull_adv_sync_reset_finalize(void)
+{
+	struct lll_adv_sync *lll_sync;
+	struct ll_adv_set *adv;
+	uint8_t handle;
+	int err;
+
+	for (handle = 0U; handle < BT_CTLR_ADV_SET; handle++) {
+		adv = ull_adv_is_created_get(handle);
+		if (!adv) {
+			continue;
+		}
+
+		lll_sync = adv->lll.sync;
+		if (!lll_sync) {
+			continue;
+		}
+
+		lll_adv_data_reset(&lll_sync->data);
+		adv->lll.sync = NULL;
+	}
```
```suggestion
```

ppryga replied:
Referring to comment: https://github.com/zephyrproject-rtos/zephyr/pull/37333#discussion_r680801307

I suggest to do not remove the function. Just change it to set all adv->lll.sync to NULL.
So this part may be removed:
```
lll_sync = adv->lll.sync;
if (!lll_sync) {
	continue;
}
lll_adv_data_reset(&lll_sync->data);
```

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -124,6 +124,25 @@ int ull_df_reset(void)
+			continue;
+		}
+
+		if (adv->df_cfg) {
+			adv->df_cfg->is_enabled = 0U;
+			ull_df_adv_cfg_release(adv->df_cfg);
+			adv->df_cfg = NULL;
+		}
```
Set to `NULL` is sufficient as `lll_df_adv_cfg_pool` is `mem_init` in `init_reset` function.

```suggestion
		adv->df_cfg = NULL;
```

ppryga replied:
Unnecessary code removed

ppryga-nordic committed:
Bluetooth: controller: Add init and reset of LLL per adv in lll_adv.c

There were missing calls to periodic advertising functions:
lll_adv_sync_init() and lll_adv_sync_reset() corresponding
functions in lll_adv.c.

Lack of those calls didn't introduce any harm.
I've added them to have complete init and reset patch,
similar to extended advertising implementation.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
@cvinayak I've refactored code to follow sequence of reset:
- reset ULL context
- reset LLL context
- reset and initialize LLL static variables, performed by `perform_lll_reset`
- reset ULL static variables,
- init ULL context variables.

ppryga commented:
@cvinayak take a look on the PR please.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1506,14 +1495,34 @@ int ull_adv_reset_finalize(void)
 	uint8_t handle;
 	int err;
 
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
+#if defined(CONFIG_BT_CTLR_ADV_AUX_SET)
+	if (CONFIG_BT_CTLR_ADV_AUX_SET > 0) {
+		err = ull_adv_aux_reset_finalize();
+		if (err) {
+			return err;
+		}
+	}
+#if defined(CONFIG_BT_CTLR_ADV_PERIODIC)
+	err = ull_adv_sync_reset_finalize();
+	if (err) {
+		return err;
+	}
+#endif /* CONFIG_BT_CTLR_ADV_PERIODIC */
+#endif /* CONFIG_BT_CTLR_ADV_AUX_SET */
+#endif /* CONFIG_BT_CTLR_ADV_EXT */
+
```
```suggestion
#if defined(CONFIG_BT_CTLR_ADV_EXT)
#if defined(CONFIG_BT_CTLR_ADV_AUX_SET)
	if (CONFIG_BT_CTLR_ADV_AUX_SET > 0) {
		err = ull_adv_aux_reset_finalize();
		if (err) {
			return err;
		}
#if defined(CONFIG_BT_CTLR_ADV_PERIODIC)
		err = ull_adv_sync_reset_finalize();
		if (err) {
			return err;
		}
#endif /* CONFIG_BT_CTLR_ADV_PERIODIC */
	}
#endif /* CONFIG_BT_CTLR_ADV_AUX_SET */
#endif /* CONFIG_BT_CTLR_ADV_EXT */

```

ppryga replied:
Change applied.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1506,14 +1495,34 @@ int ull_adv_reset_finalize(void)
 		adv->is_created = 0;
 		lll->aux = NULL;
-#endif
+#endif /* CONFIG_BT_CTLR_ADV_EXT */
+#if defined(CONFIG_BT_CTLR_ADV_PERIODIC)
+		lll->sync = NULL;
+#endif /* CONFIG_BT_CTLR_ADV_PERIODIC */
```
```suggestion
#if defined(CONFIG_BT_CTLR_ADV_PERIODIC)
		lll->sync = NULL;
#endif /* CONFIG_BT_CTLR_ADV_PERIODIC */
#endif /* CONFIG_BT_CTLR_ADV_EXT */
```

ppryga replied:
Change applied.

cvinayak reviewed (changes requested):
Minor conditional compilation changes

ppryga commented:
@cvinayak PR updated.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -434,7 +434,7 @@ int ull_adv_aux_init(void)
 }
 
-int ull_adv_aux_reset(void)
+int ull_adv_aux_reset_finalize(void)
```
I am not sure I understand this renaming: Without a `ull_adv_aux_reset`, or a `ull_adv_aux_reset_start` or anything like that, what purpose does the qualifier `finalize` have? 

I see that we have a `ull_adv_reset_finalize` but we also have a `ull_adv_reset` which doesn't seem to be the case for  `ull_adv_aux`

ppryga replied:
`ull_adv_reset` is responsible for stopping all scheduling in ULL and request to stop scheduling LLL.
After that myflay is used to call LLL to reset all statics and globals.
Then `ull_adv_reset_finalize` enters to reset and initialize ULL globals.

I didn't change the naming conventions, just change other functions that actually do ULL memory clear and initialization.

Thalley reviewed (approved):
A question regarding naming, otherwise LGTM
